high altitude description of disputes (#2285)
paritytech/polkadot: roadmap/implementers-guide/src/runtime/disputes2.md

# Disputes

## Motivation

If the transaction can be verified the defined verification code, it is good. If not, there is a dispute
about a block, which made it through the backing stage, but was then during secondary checking found

> to be invalid.

Reviewer: It's technically possible two backers disagree too.  We should not waste too much time on this special case because a backing voter could create the dispute exactly like an approval checker can, but then availability should run concurrently with the dispute, and the disputing backer should continue handing out pieces, just like if they were a backing vote.  

An incorrect disputer need not be slashed 100% but they should be slashed enough that censoring a parachain by disputing blocks you dislike becomes prohibitively expensive, and correlated slashing rule apply to drive this upwards towards 100%.

At this point the validator set can not be trusted (since they voted for the block despite something being
fishy at the very least) or the validator that blew the whistle has ulterior motives to do so.
In either way, there are malicious validators around.
As a consequence, all validators at the time of block backing, are being notified via broadcast of the first challenging vote.
Validators that backed the candidate implicitly voted for backing the candidate and may not vote a second time.

> That would be equal to an attempted double vote and would be slashed accordingly.

Reviewer: It's optional if we count their second vote, but we cannot permit later votes to replace their earlier votes of any voting type (dispute, approval, or backing), so it's almost surely simplest to allow voters to disagree with themselves.

Author: But why would we allow a voter to contradict themselves? Isn't that already bad behaviour? If we would accept multiple votes of a validator, that would influence their weight. The simplest approach would be, to allow exactly one kind of vote per validator, and silently ignore vote types that are on the same side per validator (multiple plus votes are ok, but only the first one is accounted for), voting in differently with different vote types should be an attempted double vote.

Reviewer: We cannot ignore votes if we cannot know which vote came first.  As I said below, we'll just count all votes because this harms the guy who votes multiple ways even without special handling.

Author: How do we deal with multple votes that are both pro (backing + explicit pro vote), where one comes during the open dispute, but the second comes in after dispute resolution. Then we would need to either reward the full or the reduced amount, imho it's the earlier.

## Initiation

A dispute is initiated by one secondary checker creating and gossiping a vote, that challenges the vote.
Secondary checkers run the validation code and use this to verify the transactions.

After a secondary checker challenged a block, all validators that received the gossiped vote, check the

> block themselves via the validation code.

Reviewer: "check the block" could "reconstruct and check the block" but maybe that's implicit based upon everything else

vote is concluded.
Conclusion implies that the result for this block can not be altered anymore, valid or invalid is fixed now.

In order to assure, the dispute result is not forgotten. It has to be recorded on chain.
This on chain recording mechanic must be vigilant, in a sense, that new emerging forks
must also receive the dispute resolution recorded (transplantation) iff the disputed block is container there or is in the
ancestry path.

If the disputed block was already finalized, the block must be put in governance mode to be resolved be by hand
(i.e. sudo or motion or other mechanics that are available ).

Now there might be the case that the network is fragmented for a limited amount of time. If the dispute would happen
during this time, a block could be disputed, that we have never seen being part of any chain, but we might get
a vote for it. That is called a "remote" dispute.

As such the validator has to keep track of all votes irrespective if the disputed block is already known or not.
All backing votes should be either kept in storage as well, or be queried on demand, since they are a kind of vote

> as well.

Reviewer: I'd say the dispute should contain enough data to prove that all validators should begin processing the dispute, meaning at least one backing vote and at least one disputing vote.  

We're resolving using the validator set at inclusion time, so the backing vote could actually come from another validator set, but that's okay because the backers' stake remains locked and they remain liable.

Author: What would happen if noone is backing the proposed block?

Reviewer: Good point!  If you receive an invalid candidate from a collator then what should you do?   It's not a dispute because no validator voted for the candidate yet.  It's almost surely buggy parachain code or buggy collator deployment, but maybe someone has launched an attack and just sent the bad candidate to you by mistake?  

There are two solutions, either (a) we hang onto the candidate until another candidate from that parachain gets finalized, or else (b) we throw it away and figure we can reconstruct ourselves if required, or else (c) ignore the problem and leave it to approval checkers.

I suspect Rob might dislike (a) so maybe we could implement (b) saying "backers have the right to reconstruct a block like tranche zero approval checkers".  In this way, you simply ignore invalid blocks that nobody else backs when backing, but then recheck under this bizarre situation.  We then later partially do (a) by saying that we deprioritize storing the block but do not actually flush out its storage yet, but this adds nothing formally. 

In the short term, I'd vote (c) ignore unbacked invalid blocks unless they're backed, but later maybe we'll try (b) or even this partial (a).

Reviewer 2: I think we've got to accept disputes on anything that at least one validator has issued a backing statement for. Otherwise we'd have to mess around with trying to prove to some remote chain whether a block was actually backed by a majority of its backing validators, and who those validators were, etc.

Reviewer 2: Well, now that I think about it, proving this to a remote chain might not actually be that hard assuming that every chain has the same view of sessions / validators / paras. But that is kind of a stretch as we don't fix paras one session in advance the same way we do validators/randomness. It is a change I am open to, though.

Author: But the implications of noone backing that particular block would be no inclusion, as such why would we care to deal with it any further down the row, instead of just dropping the negative vote? I think we need to talk some more about the details of that.

Reviewer: Actually my (b) sounds pointless now, so our approach goes roughly:

We've a backer who obtains a candidate B but discovers B to be invalid.  Initially, our backer does nothing but ideally^2 retains B for a little while.  If our backer ever sees a backing vote for B then our backer should ideally^1 create a dispute, and retain B permanently if they have B.  Ideally^3 our disputing backer should then participate in erasure coded piece distribution.  Ideally^4 if a backer appears absent then we ask for one more approval checker, but do not count the missing backer as a no-show.

We've three simplifications obtained by removing bits of functionality prefixed by ideally^i for i>k.  All four ideally^i add only small-ish improvements so I do not want to push for them when people have higher priority things.  If an ideally^i looks really easy to do, then go for it.  :)  If we add none of these ideally^i then backers simply ignore invalid candidates.

Author: Ideally^1 & Ideally^2 seem reasonable to impl, not sure about ^3 and ^4

# Disputes

## Motivation

> If the transaction can be verified the defined verification code, it is good. If not, there is a dispute

Reviewer 2: It's not clear what "the transaction" is here.

Reviewer 2: Also, there is no dispute if everyone agrees the block is bad.

about a block, which was either detected during the backing stage by a backing checker
or, made it through the backing stage, but was then found invalid by an approval checker.

> (If there was no conflict between votes, but nobody backed the block in the first place, it is TBD if that falls in the category of a dispute or not)

Reviewer 2: dispute implies conflict IMO. If there's nobody to slash because a backer just rejected some bad collator's block, then we have nobody to slash anyway.

> At this point the validator set can not be trusted (since they voted for the block despite something being

Reviewer 2: The validator set needs to be trusted but the backing validators may not be.

fishy at the very least) or the validator that blew the whistle has ulterior motives to do so (i.e. it is controlled by a third party and wants to incur damage to itself).
In either way, there are malicious validators around.
As a consequence, all validators at the time of block backing, are being notified via broadcast of
the first challenging vote.
Validators that backed the candidate implicitly voted for backing the candidate and may not vote a second time in different fashion, i.e. a pro and a con vote would be punished, two positive votes would be fine, but the first would be counted.
Two opposing votes by the same validator would be equal to an attempted double vote and would be slashed accordingly.

All validators at block inclusion time are eligible to (and should) cast their Vote. The backing votes of backing checkers
are counted as votes as well.

## Initiation

A dispute is initiated by one approval checker creating and gossiping a vote, that challenges the vote.
approval checkers run the validation code and use this to verify the transactions.

> After a approval checker challenged a block, all validators that received the gossiped vote, reconstruct the block

Reviewer 2: ```suggestion
After an approval checker challenged a block, all validators that received the gossiped vote, reconstruct the block
```

Generally I think we should be circulating a backing vote with it.

Author: How do we pick the backing vote? Or should all backing votes be circulated?

from availability erasure code chunks and check the block themselves via the validation code.
The result of that check is converted into a vote, and distributed via the same mechanics as the first one.

Once a receiver receives quorum and supermajority in one or the other direction, the
vote is concluded.
Conclusion implies that the result for this block can not be altered anymore, valid or invalid is fixed now.

> In order to assure, the dispute result is not forgotten or intentionally side stepped, it has to be recorded on chain.

Reviewer 2: ```suggestion
In order to ensure, the dispute result is not forgotten or intentionally side stepped, it has to be recorded on chain.
```

This on chain recording mechanic must be vigilant, in a sense, that new emerging forks
must also receive the dispute resolution recorded (transplantation) iff the disputed block is in the
ancestry path of that chain.

If the disputed block was already finalized, the block must be put in governance mode to be resolved be by hand
(i.e. sudo or motion or other mechanics that are available ).

As such the validator has to keep track of all votes irrespective if the disputed block is already known or not.
All backing votes should be either kept in storage as well, or be queried on demand, since they are a kind of vote
as well.

## Late votes

Late votes, after the dispute already reached a quorum + super majority, must be rewarded (albeit a smaller amount) as well.
These ones must be attached to the votes after a defined period of time after the result has reached
the required quorum + supermajority.

## Chain Selection / Grandpa

> Chain selection should be influenced by the chance of picking a chain that does not even include the disputed block.

Reviewer 2: I think we should avoid voting to finalize any chains that have an active dispute or have had a dispute resolve to the negative. This is combined with the voting rule of not finalizing anything which doesn't appear to have been approved by enough approval checkers.

Author: Not quite sure this section fits in here, or if that should be covered in a document itself and just be linked. It seems a bit out of scope for disputes and not affect the actual logic at all.

# Disputes

## Motivation

A reconstructed PoV can be verified with the defined verification code, that is valid during the session the block was included.
If the block is invalid and there exists at least one backing vote and one disputing vote, a dispute exists,
which was either detected during the backing stage by a backing checker
or, made it through the backing stage, but was then found invalid by an approval checker which
propagate his vote.

At this point the set of backing validators can not be trusted (since they voted for the block despite something being
fishy at the very least). On the other hand, one must also consider, the validator that blew the whistle has ulterior motives
to do so (i.e. it is controlled by a third party and wants to incur damage to itself).
In either way, there are malicious validators around.
As a consequence, all validators at the time of block backing, are being notified via broadcast of
the first challenging vote.
Validators that backed the candidate implicitly count as votes. Those validators are allowed to cast
a regular vote (a non-backing vote) as well, but it is generally not in their interest to vote both sides, since that would

> advance the progress towards super majority either way and have their bonds slashed.

Reviewer 3: It would be helpful to define the terms `quorum` and `supermajority`.
```suggestion
advance the progress towards supermajority either way and have their bonds slashed.
```

ancestry path of that chain.

If the disputed block was already finalized, the block must be put in governance mode to be resolved be by hand
(i.e. sudo or motion or other mechanics that are available ).

As such the validator has to keep track of all votes irrespective if the disputed block is already known or not.
All backing votes should be either kept in storage as well, or be queried on demand, since they are a kind of vote
as well.

## Late votes

Late votes, after the dispute already reached a quorum + super majority, must be rewarded (albeit a smaller amount) as well.
These ones must be attached to the votes after a defined period of time after the result has reached
the required quorum + supermajority.

## Chain Selection / Grandpa

Chain selection should be influenced by the chance of picking a chain that does not even include the disputed block.
Hence removing the need to include the dispute resolution itself.
This is only possible though, if the set of active heads contains such a fork.
In Grandpa the Voting rule should be used to avoid finalizing chains that contain an open or negative shut (shut with super majority that marks the block as invalid) dispute.

> In case all possible chains contains such a dispute, a TBD metric must be used to decide which fork to use or avoid finalization until one dispute resolves positive (the

Reviewer 3: TBD?

Author: To be decided, it's a bit out of scope for this document as well, since the logic of dispute resolution should not be influenced by the fork-choice rule.

about a block, which was either detected during the backing stage by a backing checker
or, made it through the backing stage, but was then found invalid by an approval checker.

(If there was no conflict between votes, but nobody backed the block in the first place, it is TBD if that falls in the category of a dispute or not)

At this point the validator set can not be trusted (since they voted for the block despite something being
fishy at the very least) or the validator that blew the whistle has ulterior motives to do so (i.e. it is controlled by a third party and wants to incur damage to itself).
In either way, there are malicious validators around.
As a consequence, all validators at the time of block backing, are being notified via broadcast of
the first challenging vote.
Validators that backed the candidate implicitly voted for backing the candidate and may not vote a second time in different fashion, i.e. a pro and a con vote would be punished, two positive votes would be fine, but the first would be counted.

> Two opposing votes by the same validator would be equal to an attempted double vote and would be slashed accordingly.

Reviewer: A priori, we do not require special punishment here.  Imagine someone votes both ways:  If the block turns out valid then they're slashed maybe 10%.  If otoh the block turns out invalid then they're slashed 100%.  If the block turns out indecisive, meaning no supermajority either way, then we lack the supermajority for any slash.  It's never in their interest to vote both ways because doing so only makes it more likely to reach a supermajority

Author: This works due to our base assumption that less than ⅓ of validators are malicious, which implies if all malicious validators vote both ways, we can never reach a supermajority (or multiple supermajorities) in any case.
